### ISSUE TITLE
SBFspotUploadDeamon not uploading

### ISSUE THREAD

**Issue by SBFspot** [SBFspot/SBFspot#154]
Hi,

After upgrading to Ubuntu 16.04 I can not get SBFspotUploadDeamon to start uploading.
If I start the deamon manually I get only:
[13:08:40] INFO: Starting SBFspotUploadDeamon Version 1.0.1
[13:08:40] INFO: Starting Daemon...

in the log, if I reboot the server I get:
[13:13:09] INFO: Starting SBFspotUploadDeamon Version 1.0.1
[13:13:09] ERROR: Unable to open database. Check configuration.

I'm using Mysql: Server version: 5.7.12-0ubuntu1 (Ubuntu)
The database gets populated correctly.

When I start the deamon the mysql query log on (turned it on temporarily) it showed:
SELECT DATE_FORMAT(Timestamp,'%Y%m%d,%H:%i'),V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12 FROM vwPvoData WHERE TimeStamp>NOW()-INTERVAL 13 DAY AND PVoutput IS NULL AND Serial=2120110081 ORDER BY TimeStamp LIMIT 30

When I tried this query in PHPmyadmin I stumbled into this problem:
1055 - Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'vwConsumption.Timestamp' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by

The deamon started uploading after I removed the 'only_full_group_by' from sql_mode. 

BR,
Hubert

#### This work item was migrated from CodePlex
CodePlex work item ID: '150'
Vote count: '1'

**Comment by SBFspot** [comment]
[hhoef@29/04/2016]
This view is, I think, the main problem. If you GROUP BY on Nearest5Min then Mysql can not witch Timestamp it should return.

CREATE VIEW vwAvgConsumption AS
SELECT Timestamp,
Nearest5min,
cast(avg(EnergyUsed) As decimal(9)) As EnergyUsed,
cast(avg(PowerUsed) As decimal(9)) As PowerUsed
FROM vwConsumption
GROUP BY Nearest5Min; 

Hope someone can fix this.

BR,
Hubert

**Comment by SBFspot** [comment]
[SBF@2/05/2016]
Timestamp should not be in the group by select. This is indeed a bug.

**Comment by SBFspot** [comment]
[SBF@2/05/2016]
Here's the fix (for MySQL only)
Two views should be altered: __vwAvgConsumption__ and __vwSpotData__ 

``` SQL
USE sbfspot;

DROP VIEW IF EXISTS vwAvgConsumption;

CREATE VIEW vwAvgConsumption AS
	SELECT Nearest5min,
	cast(avg(EnergyUsed) As decimal(9)) As EnergyUsed,
	cast(avg(PowerUsed) As decimal(9)) As PowerUsed
	FROM vwConsumption
	GROUP BY Nearest5Min;


DROP VIEW IF EXISTS vwSpotData;

CREATE View vwSpotData AS
	Select From_UnixTime(Dat.TimeStamp) AS TimeStamp,
	From_UnixTime(CASE WHEN (Dat.TimeStamp % 300) < 150
	THEN Dat.TimeStamp - (Dat.TimeStamp % 300)
	ELSE Dat.TimeStamp - (Dat.TimeStamp % 300) + 300
	END) AS Nearest5min,
	Inv.Name,
	Inv.Type,
	Dat.Serial,
	Pdc1, Pdc2,
	Idc1, Idc2,
	Udc1, Udc2,
	Pac1, Pac2, Pac3,
	Iac1, Iac2, Iac3,
	Uac1, Uac2, Uac3,
	Pdc1+Pdc2 AS PdcTot,
	Pac1+Pac2+Pac3 AS PacTot,
	CASE WHEN Pdc1+Pdc2 = 0 THEN
	    0
	ELSE
	    CASE WHEN Pdc1+Pdc2>Pac1+Pac2+Pac3 THEN
	        ROUND((Pac1+Pac2+Pac3)/(Pdc1+Pdc2)*100,1)
	    ELSE
	        100.0
	    END
	END AS Efficiency,
	Dat.EToday,
	Dat.ETotal,
	Frequency,
	Dat.OperatingTime,
	Dat.FeedInTime,
	ROUND(BT_Signal,1) AS BT_Signal,
	Dat.Status,
	Dat.GridRelay,
	ROUND(Dat.Temperature,1) AS Temperature
	FROM SpotData Dat
INNER JOIN Inverters Inv ON Dat.Serial=Inv.Serial;
```

**Comment by SBFspot** [comment]
[hhoef@2/05/2016]
I have modified the two views and the errors in phpMyadmin are gone (removed the 'only_full_group_by' exception).

Thanks for the quick fix.

BR,
Hubert

**Comment by SBFspot** [comment]
[humke@17/05/2016]
I am having the same issues but I am using sqlite3. I tried to alter some of the views based on above, but without any luck (successfully altered, but no uploads are being done). I think there also might be some missing 'AS' keywoards in the sqlite views, but could also be me not being an SQL exper :)

For example:

```
CREATE View vwSpotData AS
SELECT datetime(Dat.TimeStamp, 'unixepoch', 'localtime') TimeStamp,
        datetime(CASE WHEN (Dat.TimeStamp % 300) < 150
        THEN Dat.TimeStamp - (Dat.TimeStamp % 300)
        ELSE Dat.TimeStamp - (Dat.TimeStamp % 300) + 300
        END, 'unixepoch', 'localtime') AS Nearest5min,
        Inv.Name,
... etcetera...
```
Shouldn't the 2nd line end with 'AS TimeStamp,' ?

Would be great if someone could provide an updated CreateSQLiteDB.sql and/or CreateMySQLDB.sql that contains the correct views. That can then be used to 'diff' the needed changes.

Humke

**Comment by SBFspot** [comment]
[SBF@17/05/2016]
Don't touch it if you're not sure ;-)
The views are ok; 'AS' is optional
Did it upload before or not?

**Comment by SBFspot** [comment]
[humke@17/05/2016]
For me the upload daemon stopped working after installing 3.1.5. Before it worked fine (which I assume was version 3.1.2). Running on rpi 2 with sqlite3 as database. 

I verified that spot data is still being added to the database, but upload daemon doesn't get it out. Will try to recreate the views based on 3.1.5 'CreateSQLiteDB.sql' contents and see what happens, but that actually would be my starting point anyway when trouble began right? Or has there been changes in views from 3.1.2 to 3.1.5?

Can I use the contents of the .sql file (dropping views and create them based on .sql file contents) and should or should I not make the 'group by' changes as mentioned earlier in this thread when using sqlite3?

**Comment by SBFspot** [comment]
[SBF@18/05/2016]
The SQLite views aren't changed between 312 and 315
What's in the upload logfile?
What's the output of (will take some time on rpi)
```SQL
SELECT * FROM [vwPvoData] where pvoutput is null order by timestamp desc;
```

**Comment by SBFspot** [comment]
[humke@18/05/2016]
The latest created logfile contains the following:

```
~/slimmemeter/smadata/logs $ cat SBFspotUpload20160515.log
[14:35:45] INFO: Stopping Daemon...
[14:35:49] INFO: Starting SBFspotUploadDeamon Version 1.0.1
[14:35:49] INFO: Starting Daemon...
```

UploadDaemon is running:

```
~/slimmemeter/smadata/logs $ ps -ef | grep Upload
root       629     1  8 May15 ?        06:22:29 /usr/local/bin/sbfspot.3/SBFspotUploadDaemon

```

No output from the query (took a few seconds before sqlite> prompt returned):

```
~/slimmemeter/smadata $ sqlite3 SBFspot.db
SQLite version 3.8.7.1 2014-10-29 13:59:56
Enter ".help" for usage hints.
sqlite> SELECT * FROM [vwPvoData] where pvoutput is null order by timestamp desc;
sqlite>

```

**Comment by SBFspot** [comment]
[humke@18/05/2016]
Latest record in vwPvoData is latest record in pvoutput.org but I guess that is expected. What query is UploadDaemon running to get data?

**Comment by SBFspot** [comment]
[SBF@24/05/2016]
It seems your SBFspot isn't running every 5 mins.
When "SELECT * FROM [vwPvoData] where pvoutput is null order by timestamp desc;" doesn't return data well... there's no data to upload.
(Sorry for the late reply, but I only received the notification mails today)

**Comment by SBFspot** [comment]
[humke@25/05/2016]
SBFspot is executed every 1 minute from crontab which contains the following line:

```
# get data from Sunny Boy inverter
* 5-23 * * * /home/arjan/slimmemeter/sunnyboy_read.sh >/dev/null
```

The 'sunnyboy_read.sh' script that is executed contains the following line:

```
/usr/local/bin/sbfspot.3/SBFspot -q -ad0 -am0 -ae0 -nocsv
```
(only interested in spot data)

Doing some query like for example:
```
SELECT * FROM vwAvgSpotData;
```

returns data from today, yesterday, the day before that, etc. etc. which tells me that SBFspot is executing OK from crontab. Same type of query on 'vwSpotData' also gives results for 'every day' until today. 

Only the vwPvoData has no data after the moment things stopped working. 

But... as mentioned earlier in this thread, I did change some views in the database. Would it be a good idea to change all the views back to original and troubleshoot from there, or would you like to try/check other things first?

**Comment by SBFspot** [comment]
[SBF@25/05/2016]
Your sunnyboy_read.sh in not OK - You need daydata as this is uploaded to PVO
So, change it to -ad1
```
/usr/local/bin/sbfspot.3/SBFspot -q -ad1 -am0 -ae0 -nocsv
```
And yes, change the views back to initial state

**Comment by SBFspot** [comment]
[humke@25/05/2016]
Changing that one option does wonders:

```
[13:47:36] INFO: Uploading 90 datapoints, starting with 20160525,06:20,10763184,0,,,26.18,230.336 => OK (200)
[13:49:36] INFO: Uploading datapoint: 20160525,13:50,10766880,660,,,34.65,230.795 => OK (200)
[13:54:36] INFO: Uploading datapoint: 20160525,13:55,10766926,552,,,34.7,229.335 => OK (200)
[13:59:36] INFO: Uploading datapoint: 20160525,14:00,10766972,552,,,34.65,228.51 => OK (200)
```

So, it now looks like I changed more than just updating to 3.1.5 at the time, but I don't remember doing that. To be sure I will still change back SQL views to original (which I have not done yet) and share the results here.

First let's query the converter for (most of) the missing day data :)

**Comment by SBFspot** [comment]
[humke@25/05/2016]
Retrieving the missing day data worked fine for me by using option -ad34
What would be the difference with the option -startdate:YYYYMMDD and filling in the correct startdate?

**Comment by SBFspot** [comment]
[SBF@25/05/2016]
-ad1 gets the data of today
-ad1 -startdate:20160501 gets the data of 01-may-2016 (ideal when you have a gap somewhere and you don't want to retrieve all the data you already have

Version 4 will do this automatically ;-)

**Comment by SBFspot** [comment]
[humke@2/06/2016]
As I promised to give feedback after changing back the views to original (as they are in the 3.1.5 'CreateSQLiteDB.sql') I now am a little bit embarrassed... Everything keeps working so this pretty much proves I actually changed the option as discussed in this thread during my upgrade to 3.1.5 and that messed up my upload to pvoutput (and also getting the spotdata into an rrdtool database). 

I appreciate the support given and at least I got some more knowledge of the workings of SBFspot and the UploadDaemon in the proces. Thanks!

Now, let's download 3.3.1 and nothing else but upgrade to that version ;)
